Python/pylifecycle.c

```diff
@@ -302,6 +302,31 @@ import_init(PyInterpreterState *interp, PyObject *sysmod)
 }
 
 
+#ifdef PY_WARN_ON_C_LOCALE
+static const char *_C_LOCALE_WARNING =
+    "Python runtime initialized with LC_CTYPE=C (a locale with default ASCII "
+    "encoding), which may cause Unicode compatibility problems. Using C.UTF-8, "
+    "C.utf8, or UTF-8 (if available) as alternative Unicode-compatible "
+    "locales is recommended.\n";
+
+static void
+_emit_stderr_warning_for_c_locale(void)
+{
+    const char *coerce_c_locale = getenv("PYTHONCOERCECLOCALE");
+    /* We don't emit a warning if locale coercion has been explicitly disabled.
+     *
+     * For consistency with the corresponding check in Programs/python.c
+     * we ignore the Python -E and -I flags here.
+     */
+    if (coerce_c_locale == NULL || strncmp(coerce_c_locale, "0", 2) != 0) {
```

Comment:
The downside of that approach would be that it means embedding applications can't use `PYTHONCOERCECLOCALE=0` to silence the warning without changing the configured locale, and I don't want to define a new public embedding configuration API just to avoid checking the environment variable twice.

So I'll unconditionally export a private helper function from pylifecycle.c (so the ABI doesn't depend on the configure flag setting), and re-use that from the CLI implementation.